### ISSUE TITLE
[RFE] Let the builder decide how to handle resuming

### ISSUE THREAD

**Issue by khardix** [sclorg/rpm-list-builder#32]
## The problem

Since Koji does not allow building the same NVR multiple times, we need to adjust the packages to work around that. The solution used before was to add `.bs{num}` part into the release part of NVR for all bootstrapping packages, statring at `.bs0` and increasing as needed.

In order to do this automatically, the recipe needs to be analyzed before build, both when running from start and when resuming the build. However, the `BaseBuilder.before` method is unconditionally skipped if the build is resumed.

## Suggested change

Do not skip the `before` call. Instead, change the `BaseBuilder.before` signature to include a `resume` parameter, and let the derived builder to decide what to do when the build is resumed.

```python
def before(self, resume: int = 0, **kwargs):
    """Perform pre-build preparation.

    Keyword arguments:
        resume: Which package will the build start on.
            Defaults to 0, which means build from start.
    """
```

**Comment by khardix** [comment]
As a side note, I think that assuming "resume 0" == "build from start" could be applied globally and would be simpler (and no less correct) that the current handling.

**Comment by junaruga** [comment]
What is RFE?

`BaseBuilder.before` is intended to manage "before" action for entire build such as cleaning repository. `BaseBuilder.prepare` is intended to manage "before" action for each packages. RPM List builder is anlyzing recipe file and editing spec file there. Implementing your change at `BaseBuilder.prepare` is problem?

> I think that assuming "resume 0" == "build from start" could be applied globally and would be simpler (and no less correct) that the current handling.

I could not understand your meaning.

**Comment by junaruga** [comment]
I think you do not have to parse manually.
`rpmdev-bumpspec` command is useful isn't it?

```
$ rpmdev-bumpspec -s bs -c 'Enable bootstrap.' rubygem-actionpack.spec 
$ git diff
diff --git a/rubygem-actionpack.spec b/rubygem-actionpack.spec
index d72107b..79791c2 100644
--- a/rubygem-actionpack.spec
+++ b/rubygem-actionpack.spec
@@ -7,7 +7,7 @@
 Name: rubygem-%{gem_name}
 Epoch: 1
 Version: 5.0.1
-Release: 2%{?dist}
+Release: 2%{?dist}.bs1
 Summary: Web-flow and rendering framework putting the VC in MVC (part of Rails)
 Group: Development/Languages
 License: MIT
@@ -93,6 +93,9 @@ popd
 %doc %{gem_instdir}/README.rdoc
 
 %changelog
+* Wed May 31 2017 Jun Aruga <jaruga@redhat.com> - 1:5.0.1-2.bs1
+- Enable bootstrap.
+
 * Tue Jan 03 2017 Pavel Valena <pvalena@redhat.com> - 1:5.0.1-2
 - Enable tests.
```

**Comment by khardix** [comment]
@junaruga:

RFE - Request For Enhancement

***

The problem is that the `.bs{num}` is supposed to be applied to all builds *except* the last one. For example:

1. foo-1.0-1.bs1.el7
2. foo-1.0-1.bs2.el7
3. foo-1.0-1.el7

Each of the NVRs in that sequence is newer than the previous one.

> Implementing your change at `BaseBuilder.prepare` is problem?

In order to determine which build is last, I need to see the whole recipe. `BaseBuilder.prepare` has access only to a single package dictionary and do not see the recipe as a whole.

> `rpmdev-bumpspec` command is useful isn't it?

It is, but suffixing the `.bs{num}` breaks the NVR comparison:

```sh
$ rpmdev-vercmp 0:1.0-1.el7 0:1.0-1.el7.bs1
0:1.0-1.el7 < 0:1.0-1.el7.bs1  # bad
$ rpmdev-vercmp 0:1.0-1.el7 0:1.0-1.bs1.el7
0:1.0-1.el7 > 0:1.0-1.bs1.el7  # ok
```

Even if we were OK with the `.bs` being suffix, we still need to bump the release of the last build *without* it, so I still need to know which build of a single package is the last one.

***

>> I think that assuming "resume 0" == "build from start" could be applied globally and would be simpler (and no less correct) that the current handling.
> 
> I could not understand your meaning.

I meant we can basically just drop the whole "is_resume" logic – just assume we are always resuming.
In case of fresh build, we are just "resuming" from the start.

In other words, `--resume 0` has the same effect as not having resume at all (as far as I know).
But this is just nitpicking not directly necessary for the problem at hand.

**Comment by junaruga** [comment]
## A logic to identify bootstrap mode

> The problem is that the .bs{num} is supposed to be applied to all builds except the last one.

How to identify if current package is built for bootstrap mode or not.

Your logic is
* if the package is last, it is not bootstrap.
* Otherwise, it is bootstrap.
If you will use your logic, you may implement it on `BaseBuilder.before`.
But actually what you want is to normalize the recipe data.

In that case, implementing at `recipe.py` looks better.

But my predicted logic is
* if the package has macro of string including "bootstrap" in`replaced_macros` or `macros`, it is bootstrap.
* Otherwise, it is not bootstrap.

So, in my case, I thought `BaseBuilder.prepare` makes sense.
It's not perfect reasonable logic. But it may be practical.

I prefer my idea. Because in the future, we want to identify bootstrap mode from only macro value ideally.


## How to edit version number for bootstrapping.

> It is, but suffixing the .bs{num} breaks the NVR comparison:

OK I see. So, you may have to parse and replace manually.
I just have idea to append "0." to the top of VERSION STRING.

```
"Release: STRING%{?dist}" => "Release: 0.STRING%{?dist}"
```

This idea is mentioned here.
https://pagure.io/koji/issue/416#comment-440659

## Dropping `is_resume`

> In other words, --resume 0 has the same effect as not having resume at all (as far as I know).

I see. I understand your meaning.
But I can not judge if I like or not until I will see the source code.

**Comment by junaruga** [comment]
## A logic to identify bootstrap mode

> The problem is that the .bs{num} is supposed to be applied to all builds except the last one.

How to identify if current package is built for bootstrap mode or not.

Your logic is
* if the package is last, it is not bootstrap.
* Otherwise, it is bootstrap.
If you will use your logic, you may implement it on `BaseBuilder.before`.
But actually what you want is to normalize the recipe data.

In that case, implementing at `recipe.py` looks better.

But my predicted logic is
* if the package has macro of string including "bootstrap" in`replaced_macros` or `macros`, it is bootstrap.
* Otherwise, it is not bootstrap.

So, in my case, I thought `BaseBuilder.prepare` makes sense.
It's not perfect reasonable logic. But it may be practical.

I prefer my idea. Because in the future, we want to identify bootstrap mode from only macro value ideally. My idea is closer to the ideal logic.
But it's up to you.

## How to edit version number for bootstrapping.

> It is, but suffixing the .bs{num} breaks the NVR comparison:

OK I see. So, you may have to parse and replace manually.
I just have idea to append "0." to the top of VERSION STRING.

```
"Release: STRING%{?dist}" => "Release: 0.STRING%{?dist}"
```

This idea is mentioned here.
https://pagure.io/koji/issue/416#comment-440659

## Dropping `is_resume`

> In other words, --resume 0 has the same effect as not having resume at all (as far as I know).

I see. I understand your meaning.
But I can not judge if I like or not until I will see the source code.

**Comment by khardix** [comment]
## A logic to identify bootstrap mode

> if the package has macro of string including "bootstrap" in `replaced_macros` or `macros`, it is bootstrap.
>  It's not perfect reasonable logic. But it may be practical.

It is practical and could be used for now, unless we find an existing recipe that breaks these assumptions. I will go through them, and if I do not find any practical issue, I will probably go with your logic.

## How to edit version number for bootstrapping

Actually, I came with a way to do it by "abusing" the `%dist` macro:

    %{expand:%global dist .bs0%dist}

This way, I have to only insert one extra line and do not have to grep for `Release` or similar. The `%expand` macro is necessary to not get infinite recursion on `%dist` redefinition.

> I just have idea to append "0." to the top of VERSION STRING.

Please keep in mind that we could have multiple bootstrapping instance of a package in the recipe, and that the simple `0.` prefix might not be enough. @hhorak, can this occur or am I too paranoid?

## Dropping `is_resume`

I will send a PR with the proposed changes and let you judge that (once the [Click CLI PR][click-pr] will be either complete or discarded).

[click-pr]: https://github.com/sclorg/rpm-list-builder/pull/21

**Comment by hhorak** [comment]
> Please keep in mind that we could have multiple bootstrapping instance of a package in the recipe, and that the simple 0. prefix might not be enough. @hhorak, can this occur or am I too paranoid?

It does not happen very often, but I saw it once or twice -- So, if that would bring too big complexity, I'd be fine with call this "a special case which requires manual intervention" and let people solve it manually.

**Comment by junaruga** [comment]
> Please keep in mind that we could have multiple bootstrapping instance of a package in the recipe, and that the simple 0. prefix might not be enough. @hhorak, can this occur or am I too paranoid?

What is "multiple bootstrapping instance"?
Why it is not enough? Could you tell me with examples?

**Comment by khardix** [comment]
@junaruga 
> What is "multiple bootstrapping instance"?

As confirmed by @hhorak, there may be more than one build of single package that should be considered bootstrap. For example:

```yaml
packages:
  - foo:  # first bootstrap build/instance
      macros:
        _with_bootstrap: 1
  - foo:  # second bootstrap build/instance
      macros:
        some_other_macro: 1
  - foo  # regular, non-bootstrap build
```

In that case, with your number prefixing, we would end with something like below (assuming the `foo-1.0-1.el7` is the final package NVR):

```
foo-1.0-0.1.el7  # first bootstrap
foo-1.0-1.1.el7  # second bootstrap; NVR "newer" than final!
foo-1.0-1.el7    # final; overriden by second bootstrap
```
```sh
$ rpmdev-vercmp 0:1.0-1.1.el7 0:1.0-1.el7
0:1.0-1.1.el7 > 0:1.0-1.el7
```

Strangely enough, the `bs0.{release}` works:

```
foo-1.0-bs0.1.el7  # first bootstrap
foo-1.0-bs1.1.el7  # second bootstrap
foo-1.0-1.el7      # final
```
```sh
$ rpmdev-vercmp 0:1.0-bs1.1.el7 0:1.0-1.el7
0:1.0-bs1.1.el7 < 0:1.0-1.el7
```

So in the end, I would go with your slightly modified idea.

```
"Release: STRING%{?dist}" => "Release: bs0.STRING%{?dist}"
```

**Comment by khardix** [comment]
## A logic to identify bootstrap mode

> > If the package has macro of string including "bootstrap" in `replaced_macros` or `macros`, it is bootstrap.
> > It's not perfect reasonable logic. But it may be practical.
>
> It is practical and could be used for now, unless we find an existing recipe that breaks these assumptions. I will go through them, and if I do not find any practical issue, I will probably go with your logic.

I just found one instance where this would break. Recipe `rh-ror50`:

```yaml
packages:
  # Packages required by Rails
  - rubygem-dalli:
      replaced_macros:
        enable_test: 0
  # Later
  - rubygem-dalli
```

There is no "bootstrap" string in `macros` or `replace_macros`, but we need build the first one in bootstrap mode.

**Comment by junaruga** [comment]
I am in vacation now.
Add enable_test as string to identify booystrap as temporary fix?

**Comment by khardix** [comment]
>  Add enable_test as string to identify booystrap as temporary fix?

Sure, but there is no guarantee some other macro will not break that again. I would prefer more systematic approach (i.e. my original solution).

> I am in vacation now.

Enjoy :) I will be on PyCon CZ for the rest of the week, so feel free to not react until you are back in office.

**Comment by junaruga** [comment]
Yeah, it's up to you. :) ok.

**Comment by hroncok** [comment]
I don't understand why would we need a special `bs` indicator in release to indicate it's a bootstrap. We don't do that in Fedora, so the concept is new to me.

**Comment by khardix** [comment]
> I don't understand why would we need a special `bs` indicator in release to indicate it's a bootstrap. We don't do that in Fedora, so the concept is new to me.

AFAIK Fedora does bootstrapping differently (using `stageX` scripts instead of SPEC files, etc.). In order to bootstrap a SCL in Koji, we first need to break circular dependencies (using macros to disable test suites and similar). Then, after the dependency circle is complete, we need to build the "bootstrapped" package again, this time with no adjustments.

The problem is as described in the issue description – Koji will not allow multiple successful builds of a package with the same NVR. Thus we need to automatically adjust that for each build, either using the indicator or similar NVR magic.

**Comment by hroncok** [comment]
Well the reason for bootsraping in Fedora is the same, for example see this comment form python3.spec:

```
# NOTES ON BOOTSTRAPING PYTHON 3.6:
#
# Due to a dependency cycle between Python, gdb, rpm, pip, setuptools, wheel,
# and other packages, in order to rebase Python 3 one has to build in the
# following order:
#
# 1. At the same time:
#     - gdb without python support (add %%global _without_python 1 on top of gdb's SPEC file)
#     - python-rpm-generators with bootstrapping_python set to 1
#       (this can be done also during step 2., but should be done before 3.)
# 2. python3 with with_rewheel set to 0
# 3. At the same time:
#     - gdb with python support (remove %%global _without_python 1 on top of gdb's SPEC file)
#     - python-rpm-generators with bootstrapping_python set to 0
#       (this can be done at any later step without negative effects)
# 4. rpm
# 5. python-setuptools with bootstrap set to 1
# 6. python-pip with build_wheel set to 0
# 7. python-wheel with %%bcond_without bootstrap
# 8. python-setuptools with bootstrap set to 0 and also with_check set to 0
# 9. python-pip with build_wheel set to 1
# 10. pyparsing
# 11. python3 with with_rewheel set to 1
#
# Then the most important packages have to be built, starting from their
# various leaf dependencies recursively. After these have been built, a
# targeted rebuild should be requested for the rest.
#
# Currently these packages are recommended to have been built before a targeted
# rebuild after a python abi change:
#   python-sphinx, pytest, python-requests, cloud-init, dnf, anaconda, abrt
```

And see those changelog entries:

```
* Fri Dec 09 2016 Charalampos Stratakis <cstratak@redhat.com> - 3.6.0-0.6.rc1
- Enable rewheel

* Wed Dec 07 2016 Charalampos Stratakis <cstratak@redhat.com> - 3.6.0-0.5.rc1
- Update to Python 3.6.0 release candidate 1
```

**Comment by khardix** [comment]
I see your point. However, if we just bump the release during SCL bootstrapping (as Fedora does), we will end with different spec and NVR than the non-SCL version has. When the non-SCL version receive an update, we would need to (probably manually) merge the changes in it atop the SCL version.

From this point of view, the `bs` indicator is an attempt at doing "invisible" bootstrapping, so that future updates could be as simple as just rebuilding the SRPM in the SCL environment.

Or at least this is my understanding of the matter. @hhorak, any comments on that?

**Comment by khardix** [comment]
After a discussion with @junaruga, we agreed that a better place for the bootstrap logic would be in the `Recipe` class, not in the builder itself. See the PR above for an implementation of that.
